src/bin/pg_autoctl/service_postgres_ctl.c

```diff
@@ -0,0 +1,516 @@
+				continue;
+			}
+			exit(EXIT_CODE_QUIT);
+		}
+
+		/*
+		 * This postgres controller process is running Postgres as a child
+		 * process and thus is responsible for calling waitpid() from time to
+		 * time.
+		 */
+		pid = waitpid(-1, &status, WNOHANG);
+
+		switch (pid)
+		{
+			case -1:
+			{
+				/* if our PostgresService stopped, just continue */
+				if (errno != ECHILD)
+				{
+					log_error("Failed to call waitpid(): %m");
+				}
+				break;
+			}
+
+			case 0:
+			{
+				/*
+				 * We're using WNOHANG, 0 means there are no stopped or exited
+				 * children, it's all good. It's the expected case when
+				 * everything is running smoothly, so enjoy and sleep for
+				 * awhile.
+				 */
+				break;
+			}
+
+			default:
+			{
+				if (pid != postgresService.pid)
+				{
+					/* might be one of our pg_controldata... */
+					char *verb = WIFEXITED(status) ? "exited" : "failed";
+					log_debug("waitpid(): process %d has %s", pid, verb);
+				}
+
+				/*
+				 * Postgres is not running anymore, the rest of the code will
+				 * handle that situation, just continue.
+				 */
+				break;
+			}
+		}
+
+		if (pg_setup_pgdata_exists(pgSetup))
+		{
+			/*
+			 * If we have a PGDATA directory, now is a good time to initialize
+			 * our LocalPostgresServer structure and its file paths to point at
+			 * the right place: we need to normalize PGDATA to its realpath
+			 * location.
+			 */
+			if (!pgStatusPathIsReady)
+			{
+				/* initialize our Postgres state file path */
+				if (!local_postgres_set_status_path(postgres, false))
+				{
+					/* highly unexpected */
+					log_error("Failed to build postgres state file pathname, "
+							  "see above for details.");
+
+					/* maybe next round will have better luck? */
+					continue;
+				}
+
+				pgStatusPathIsReady = true;
+
+				log_trace("Reading current postgres expected status from \"%s\"",
+						  localStatus->pgStatusPath);
+			}
+		}
+		else if (!pgStatusPathIsReady)
+		{
+			/*
+			 * If PGDATA doesn't exists yet, we didn't have a chance to
+			 * normalize its filename and we might be reading the wrong file
+			 * for the Postgres expected status. So we first check if our
+			 * pgSetup reflects an existing on-disk instance and if not, update
+			 * it until it does.
+			 *
+			 * The keeper init process is reponsible for running pg_ctl initdb.
+			 *
+			 * Given that we have two processes working concurrently and
+			 * deciding at the same time what's next, we need to be cautious
+			 * about race conditions. We add extra checks around existence of
+			 * files to make sure we don't get started too early.
+			 */
+			PostgresSetup newPgSetup = { 0 };
+			bool missingPgdataIsOk = true;
+			bool postgresNotRunningIsOk = true;
+
+			if (pg_setup_init(&newPgSetup,
+							  pgSetup,
+							  missingPgdataIsOk,
+							  postgresNotRunningIsOk) &&
+				pg_setup_pgdata_exists(&newPgSetup) &&
+				pg_auto_failover_default_settings_file_exists(&newPgSetup))
```

Comment:
Because we first create a buffer string with the content of the file and then write to disk in a single `write()` operation, I'm not sure that's a high-risk...

Reply:
I think that's fair, let's not care for now then.